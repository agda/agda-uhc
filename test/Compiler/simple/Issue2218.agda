
open import Agda.Builtin.Float
open import Agda.Builtin.Unit
open import Agda.Builtin.Reflection
open import Agda.Builtin.String
open import Agda.Builtin.Bool
open import Common.IO

data It (a : Float) : Set where
  it : It a

Inf : Float
Inf = primFloatDiv 1.0 0.0

-Inf : Float
-Inf = primFloatNegate Inf

NaN : Float
NaN = primFloatDiv 0.0 0.0

-NaN : Float
-NaN = primFloatNegate NaN

_==_ = primFloatEquality

macro
  literal : Float → Term → TC ⊤
  literal x hole = unify hole (lit (float x))

nl = putStrLn ""

header = putStrLn " Inf  -Inf   NaN  -NaN  |  x"

tick : Bool → String
tick true  = "   x  "
tick false = "   -  "

check : Float → IO _
check x = putStr (tick (x ==  Inf)) ,,
          putStr (tick (x == -Inf)) ,,
          putStr (tick (x ==  NaN)) ,,
          putStr (tick (x == -NaN)) ,,
          putStr "|  " ,, printFloat x ,,
          nl

main : IO _
main =
  header ,,
  check (literal  Inf) ,,
  check (literal -Inf) ,,
  check (literal  NaN) ,,
  check (literal -NaN)
