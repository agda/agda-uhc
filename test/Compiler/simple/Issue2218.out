EXECUTED_PROGRAM

ret > ExitSuccess
out >  Inf  -Inf   NaN  -NaN  |  x
out >    x     -     -     -  |  Infinity
out >    -     x     -     -  |  -Infinity
out >    -     -     x     x  |  NaN
out >    -     -     x     x  |  NaN
out >
